Fix non-griffin build for 3DS and make it the default

pstef opened the pull request:
While here:
* remove input_overlay and task_overlay from linking blacklist
* move -Werror to CC invocation since C++ refuses to use it
* copy the quiet build mode from the main Makefile and make it the default
* temporarily define DEPRECATED for gpu_old.h to allow succesful build

pstef commented:
Waiting for reviews and testing.

pstef commented:
Trying to update assets invokes kernel panic in the network stack, even on Luma 13.3.2. So I'll leave that disabled.

```Processor: Arm11 (core 1)
Exception type: prefetch abort (kernel panic)
Fault status: debug event
Current process: socket (0004013000002E02)
R0   80000000  R1    00000000
R2   00000000  R3    00000000
R2   00000007  R5    04425174
R6   00001C72  R7    1FFB1480
R8   FFF796CC  R9    FFF7B3AC
R10  00000020  R11   00000004
R12  00000000  SP    FF56DD10
LR   FFF23584  PC    FFF1C1D8
CPSR A0000013  FPEXC 00000000

Stack dump:
FF56DD: 0008 0000 0200 0000
FF56DD: 0000 0000 8088 E6E6
FF56DD: 0000 0000 0900 0000
FF56DD: CC96 F7FF 0700 0000
FF56DD: 48DD 56FF E00E 0000
FF56DD: E0DE 2608 882A F2FF
FF56DD: 8015 FB1F 0000 0000
FF56DD: 0700 0000 0000 0000
FF56DD: 8088 E6E6 0900 0000
FF56DD: DC96 F7FF 0700 0000
FF56DD: 8014 FB1F 0060 1910
FF56DD: 3000 0000 7478 F1FF
FF56DD: 0060 1910 0050 2D23
FF56DD: ACB3 F7FF 7808 F8FF
FF56DD: 8014 FB1F CC96 F7FF
FF56DD: ACB3 F7FF B096 F7FF
FF56DD: 90B3 F7FF 68C4 F7FF
FF56DD: ACB3 F7FF CC96 F7FF
FF56DD: 90B3 F7FF B096 F7FF
```

pstef commented:
I don't understand the intention behind calling command_event(CMD_EVENT_QUIT...) here: https://github.com/libretro/RetroArch/commit/9ae2514009ea8b0071bd2244a10c6847830744ea#diff-52a5a7f5fc5e26696b5747f179be9005bdc131a5776b4a73f61a3e065383e666R116
so I don't know how to replace it. If I can't replace it, this would have to be the first audio driver to include command.h

pstef commented:
This also feels to me like a hack:
```
#if defined(_3DS) && !defined(_POSIX_THREADS)
#include <../rthreads/ctr_pthread.h>
#endif
```
`deps/rcheevos/src/rc_compat.c` refers to functions from the `pthread_mutex_` family and needs their prototypes. In a griffin build those become available due to header pollution from `libretro-common/rthreads` (declaring them static is meaningless in this mode).

I got my devkitpro fairly recently and the prototypes are there, so I didn't see the problem with rc_compat.c until I saw the build failure on CI. I assume it's got an older version of the toolchain, where the functions are not implemented yet, or at least their prototypes are not exposed.

pstef commented:
With this patch I'm able to compile mbedtls into the build, but I was unable to communicate with buildbot.libretro.com over https.

pstef commented on deps/rcheevos/src/rc_compat.c:
```diff
@@ -3,6 +3,10 @@
 #include <ctype.h>
 #include <stdarg.h>
 
+#if defined(_3DS) && !defined(_POSIX_THREADS)
+#include <../rthreads/ctr_pthread.h>
+#endif
+
```
`deps/rcheevos/src/rc_compat.c` refers to functions from the `pthread_mutex_` family and needs their prototypes. In a griffin build those become available due to header pollution from `libretro-common/rthreads/rthreads.c` (declaring them static is meaningless in this mode).

I got my devkitpro fairly recently and the prototypes are there (guarded behind ifdef `_POSIX_THREADS`), so I didn't see the problem with rc_compat.c until I saw the build failure on CI. I assume it's got an older version of the toolchain, where the functions are not implemented yet, or at least their prototypes are not exposed.

I don't see a better solution to this problem that wouldn't require a slightly invasive rewrite. Except of course updating the toolchain used in the CI infrastructure.

pstef replied:
Sorry for the noise, I saw your name in git blame output without realizing what b7ddac2431f was about.

pstef commented on audio/drivers/ctr_dsp_audio.c:
```diff
@@ -113,7 +114,7 @@ static ssize_t ctr_dsp_audio_write(void *data, const void *buf, size_t len)
             if (!aptMainLoop())
             {
-               command_event(CMD_EVENT_QUIT, NULL);
+               error_and_quit("dsp_audio_write");
```
@justinweiss would you be able to help me with this?

justinweiss replied:
If aptMainLoop returns false, it means the app should quit -- not forcing a quit here would cause the app to hang, but I don't think it matters how the quit happens.

The hang might have been because the sample_pos wasn't increasing so it'd get stuck in the loop, so there could be a better way to work around this.

pstef replied:
Thanks. Would you say my idea of replacing it with error_and_quit() is sound or do you see a better solution? I understand if you don't have time for this, but in that case, I'll go with error_and_quit().

pstef commented:
If you're asking me, I was hoping for more input, especially regarding the rthreads part. If you're OK with what's here, I think we can leave it like this and possibly improve in a separate PR.

I was also looking into enabling TLS and I think I got an idea today, so that can get into this PR or I can open a separate one for it.

BTW, I know it's unrelated, but personally I would like to see #17736 in, unless there's a reason not to merge.

pstef commented:
I imported rc_compat.[ch] from upstream (arguably not ideal, should update everything) to get the changes for 3DS. 
It compiles both with the newer toolchain and the old one, but I suspect it's because it uses <3DS/synchronization.h> everywhere and not the pthread implementation that comes from devkitARM.
This also brings in changes for Windows which is arguably out of scope for this change.

LibretroAdmin commented:
Alright, should we give @MrHuu and @Jamiras like a 5 day window to comment on this? After that period expires it could then be merged

pstef commented:
Sounds good to me, thanks.

Jamiras commented:
I have no way to test the 3DS changes to rc_compat.c. Another user provided those. 

It looks like you've pulled in the chunks that were #ifdef'd for 3DS, so it seems like it should be correct. If it's working for you, I'm fine with those changes. Note that the whole file will be replaced the next time I upgrade the rcheevos dependency.

I have no comments on the rest of the changes.

MrHuu commented:
I did not build all available cores, but i got linking issues with the following cores:

```mame2003```
```mame2003_plus```
```pcsx_rearmed```

For these cores it complains about 'undefined reference to ```uncompress```'.
It's included in 'libretro-common/include/compat/zlib/zlib.h'

Otherwise, it looks good  to me.

Not sure if we need ```HAVE_AUDIOMIXER```. It does work but required menu scrolling / selecting audio files are not distributed in with the 3ds builds and have to be manually obtained. But when scrolling through a large playlist with lots of art the sound is lagging. It's probably best to remove and leave it disabled.

pstef commented:
> For these cores it complains about 'undefined reference to `uncompress`'. It's included in 'libretro-common/include/compat/zlib/zlib.h'

I added uncompr.o to Makefile.common to fix that.
 
> Not sure if we need `HAVE_AUDIOMIXER`. It does work but required menu scrolling / selecting audio files are not distributed in with the 3ds builds and have to be manually obtained. But when scrolling through a large playlist with lots of art the sound is lagging. It's probably best to remove and leave it disabled.

Thanks, I just removed it. I never needed it, only copied mindlessly from another makefile.

MrHuu commented:
Awesome! As far as i can tell it should be ready to go.
Thanks for the efforts @pstef

pstef commented:
> Not sure if we need `HAVE_AUDIOMIXER`. It does work but required menu scrolling / selecting audio files are not distributed in with the 3ds builds and have to be manually obtained. But when scrolling through a large playlist with lots of art the sound is lagging. It's probably best to remove and leave it disabled.

MrBoom references it, so I'm going to restore HAVE_AUDIOMIXER as a stop-gap before the upcoming release. And we can revisit this later.
